arch/arm/core/aarch32/cache.c

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright (c) 2013-2014 Wind River Systems, Inc.
+ * Copyright (c) 2020-2022 Qualcomm Innovation Center, Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+/**
+ * @file
+ * @brief Cache manipulation
+ *
+ * This module contains functions for manipulation caches.
+ */
+
+#include <zephyr/arch/cpu.h>
+#include <zephyr/cache.h>
+
+void arch_dcache_enable(void)
+{
+	SCB_EnableDCache();
+}
+
+void arch_dcache_disable(void)
+{
+	SCB_DisableDCache();
+}
+
+void arch_icache_enable(void)
+{
+	SCB_EnableICache();
+}
+
+void arch_icache_disable(void)
+{
+	SCB_DisableICache();
+}
+
+#ifdef CONFIG_DCACHE_LINE_SIZE_DETECT
+size_t arch_dcache_line_size_get(void)
+{
+	return __SCB_DCACHE_LINE_SIZE;
+}
+#endif /* CONFIG_DCACHE_LINE_SIZE_DETECT */
+
+#ifdef CONFIG_ICACHE_LINE_SIZE_DETECT
+size_t arch_icache_line_size_get(void)
+{
+	return __SCB_ICACHE_LINE_SIZE;
+}
+#endif /* CONFIG_ICACHE_LINE_SIZE_DETECT */
```

Comment:
Why not using `CONFIG_ICACHE_LINE_SIZE` instead if the value can be inferred at compile time? (same for dcache)

Reply:
Do you mean why not using CONFIG_ICACHE_LINE_SIZE in the build config (and having CONFIG_ICACHE_LINE_SIZE_DETECT not defined) or do you mean why using __SCB_ICACHE_LINE_SIZE in here rather than CONFIG_ICACHE_LINE_SIZE  ? 
If the former then just to provide maximum flexibility, if the latter just for consistency with the other functions in the file.

Reply:
Thanks for confirming and I completely agree with you. Just trying to keep all options open to users that was all. I'm happy to remove the arch_d/icache_line_size_get()

Reply:
> Why not using CONFIG_ICACHE_LINE_SIZE instead if the value can be inferred at compile time? (same for dcache)

But this code is *required* for dynamic cache size detection as implied by the "CONFIG_ICACHE_LINE_SIZE_DETECT" flag - so we need this, no?

Reply:
@quic-egmc  I'm not sure I got what you mean.

When the user code requires to know (for example) the d-cache line size, it must call `sys_cache_data_line_size_get()`, defined in: https://github.com/zephyrproject-rtos/zephyr/blob/48c87f2bf8b768f550cfe6756d3cd313f9a02c3f/include/zephyr/cache.h#L141-L150

Now, for architectures that have a cache line size detection at run-time this can be done defining `CONFIG_DCACHE_LINE_SIZE_DETECT` and implementing `cache_data_line_size_get()`. The architectures that have a cache line size defined at build-time (it looks like this is the case) can simply set `CONFIG_DCACHE_LINE_SIZE` and be done with that.

Reply:
> Why not using CONFIG_ICACHE_LINE_SIZE instead if the value can be inferred at compile time? (same for dcache)

Sometimes it's compile time but, in theory, for other cores it can be runtime.  

>  You are returning a constant build-time value so not sure how flexible that can be :)

I think you understandably reached this conclusion by searching and found these defines:

https://github.com/zephyrproject-rtos/cmsis/blob/5f86244bad4ad5a590e084f0e72ba7a1416c2edf/CMSIS/Core/Include/cachel1_armv7.h#L45-L51

but I think the intent behind these is that it *could* be runtime on other v7M architectures that support configurable cache sizes.  But just theoretical until ARM releases one of these cores so seems reasonable to remove this support until such time.

Reply:
>  I think the intent behind these is that it could be runtime on other v7M architectures that support configurable cache sizes.

You mean that CMSIS can change the `__SCB_DCACHE_LINE_SIZE` definition in the future so that it is not a static define anymore but some kind of macro returning a run-time detection of the line size?

I mean, they must be crazy to do something like that. I would expect that they released a new function for that instead of changing a symbol behind the scenes unexpectedly. The only thing that I can see supporting the idea of having run-time detection is the case a platform would re-define the `__SCB_DCACHE_LINE_SIZE` in some header file.

Also please note that this PR is enabling the cache management with that cache line size for **all** the AARCH32 architectures. Is that what you want? Reading the comment it seems that `__SCB_DCACHE_LINE_SIZE` is only valid for Cortex-M7.

So, I'm fine with having a run-time detection but please check whether you actually want to unconditionally enable this for all the AArch32 platforms or instead you want to restrict this to some selected SoCs.

Reply:
> You mean that CMSIS can change the `__SCB_DCACHE_LINE_SIZE` definition in the future so that it is not a static define anymore but some kind of macro returning a run-time detection of the line size?

I think it's extremely unlikely this will change for Armv7-M, and as mentioned the M7 is the only chip concerned here with a fixed 32 B line size.

The ArmV8-M chips, including the newest Cortex-M85, also have fixed 32 B line sizes:

- M85: https://developer.arm.com/documentation/101924/0002/Memory-system/Instruction-and-data-cache/L1-instruction-cache
- M55: https://developer.arm.com/documentation/101051/0002/Memory-system/Instruction-and-data-cache/L1-instruction-cache

PS: I did find some custom cache-mechanisms on M33s (V8-M), however, such as the STM32L5 which has a **16 B line size** ... but I'll have to look at the datasheet for that chip in more detail: https://www.st.com/content/ccc/resource/training/technical/product_training/group1/4c/20/41/6b/46/97/42/11/STM32L5-System-Instruction_Cache_ICACHE/files/STM32L5-System-Instruction_Cache_ICACHE.pdf/_jcr_content/translations/en.STM32L5-System-Instruction_Cache_ICACHE.pdf

Reply:
@microbuilder Can we come up with an aarch32-common solution for dynamically determining cache line size?  Or barring that one solution for ARMv7M and one for ARMv8M?

First question is architectural - is there a common location, second question is practical - does CMSIS provide and abstraction that will work across all cache-enabled implementations?

Reply:
@quic-egmc I'll try to ping the CMSIS team at Arm and see what they think about the cache line size definition longer term.

Reply:
Comment from contact at Arm: `The current Cortex Ms are fixed size so I think compile time macro makes sense in the CMSIS core header.` So seems unlikely to change in the near future for M-class devices anyway, or be configurable.

This still leaves the question of how to handle this functionality with a generic Aarch32 solution, though, across A32+R32+M.